runner/runner.py

```diff
@@ -19,46 +18,52 @@
 obligatory_plugins = [PipelineAnalyser(), IntegrityChecker()]
 
 
-def overwrite_model_configs(config: RunConfig, pipeline: Pipeline) -> Pipeline:
+def overwrite_model_configs_(config: RunConfig, pipeline: Pipeline) -> None:
     for key, value in vars(config).items():
         if value is not None:
             for model in flatten(pipeline.children()):
                 if hasattr(model, "config"):
                     if hasattr(model.config, key):
                         vars(model.config)[key] = value
 
-    return pipeline
-
-
-def add_position_to_block_names(pipeline: Pipeline) -> Pipeline:
-    entire_pipeline = pipeline.children()
-
-    def add_position(block: Union[List[Element], Element], position: int, prefix: str):
-        if isinstance(block, List):
-            if position > 0:
-                prefix += f"{position - 1}-"
-            for i, child in enumerate(block):
-                add_position(child, i, prefix)
-        elif not isinstance(block, DataSource):
-            block.id += f"{prefix}{position}"
-
-    add_position(entire_pipeline, 1, "-")
-
-    return pipeline
-
 
-def append_pipeline_id(pipeline: Pipeline) -> Pipeline:
+def append_parent_path_and_id(pipeline: Pipeline) -> None:
     entire_pipeline = pipeline.dict_children()
 
-    def append_id(block, pipeline_id: str):
-        block["obj"].pipeline_id = f"{pipeline_id}"
+    blocks_encountered = []
 
-        if "children" in block:
-            for child in block["children"]:
-                append_id(child, f"{pipeline_id}/{block['name']}")
+    def append(block, parent_path: str, id_with_prefix: str):
+        block["obj"].parent_path = f"{parent_path}"
+        block["obj"].id += f"{id_with_prefix}"
+        blocks_encountered.append(id(block["obj"]))
 
-    append_id(entire_pipeline, Const.output_pipelines_path)
-    return pipeline
+        if "children" in block:
+            for i, child in enumerate(block["children"]):
+                child_id = (
+                    f"{id_with_prefix}-{i}"
+                    if id(child["obj"]) not in blocks_encountered
+                    else f"{id_with_prefix}+{i}"
+                )
+                append(
+                    child,
+                    parent_path=f"{parent_path}/{block['obj'].id}",
+                    id_with_prefix=child_id,
+                )
+
+    append(entire_pipeline, Const.output_pipelines_path, "")
+
+
+def rename_input_id_(
+    pipeline: Pipeline, data: Dict[str, Union[pd.Series, List]]
+) -> None:
+    datasources = [
+        block
+        for block in flatten(pipeline.children())
+        if type(block) is DataSource
+        and block.id.split("-")[0].split("+")[0] == Const.input_col
```

Comment:
what's this? not sure how Const.input_col comes into the picture, as the id of the input? the input id right now is arbitrarily defined by the user.

Reply:
if this is there to differentiate between Concat and DataSource, then let's make Concat not a subclass of DataSource! It'll have the same methods anyway (and we can introduce the concept of an interface later)

Reply:
To be honest, I wanted to discuss this -- what if there are multiple input datasources all named differently? 

This is the problem I faced and I went with a quick but dirty solution:

- All blocks in the hierarchy get an ``id``, even datasources 
- But if that's the case the id referring to the input data in ``store`` is different from the new ``id`` appended to the datasource

There are two solutions to this:
- Either change the ``id`` of the input in ``store``
- or don't append the position to the datasource in the first place 

I went with ``1`` but maybe ``2`` is a better solution -- the tradeoff is that its harder to read and track where the datasource was.